packages/@aws-cdk/aws-ec2/lib/cfn-init.ts

```diff
@@ -0,0 +1,362 @@
+    const configSets = (attachOptions.configSets ?? ['default']).join(',');
+    const printLog = attachOptions.printLog ?? true;
+
+    if (attachOptions.embedFingerprint ?? true) {
+      // It just so happens that the comment char is '#' for both bash and PowerShell
+      attachOptions.userData.addCommands(`# fingerprint: ${fingerprint}`);
+    }
+
+    if (attachOptions.platform === InitPlatform.WINDOWS) {
+      const errCode = attachOptions.ignoreFailures ? '0' : '$LASTEXITCODE';
+      attachOptions.userData.addCommands(...[
+        `cfn-init.exe -v ${resourceLocator} -c ${configSets}`,
+        `cfn-signal.exe -e ${errCode} ${resourceLocator}`,
+        ...printLog ? ['type C:\\cfn\\log\\cfn-init.log'] : [],
+      ]);
+    } else {
+      const errCode = attachOptions.ignoreFailures ? '0' : '$?';
+      attachOptions.userData.addCommands(...[
+        // Run a subshell without 'errexit', so we can signal using the exit code of cfn-init
+        '(',
+        '  set +e',
+        `  /opt/aws/bin/cfn-init -v ${resourceLocator} -c ${configSets}`,
+        `  /opt/aws/bin/cfn-signal -e ${errCode} ${resourceLocator}`,
+        ...printLog ? ['  cat /var/log/cfn-init.log >&2'] : [],
+        ')',
+      ]);
+    }
+  }
+
+  private bind(scope: Construct, options: AttachInitOptions): { configData: any, authData: any } {
+    const nonEmptyConfigs = mapValues(this._configs, c => c.isEmpty ? undefined : c);
+
+    const configNameToBindResult = mapValues(nonEmptyConfigs, c => c._bind(scope, options));
+
+    return {
+      configData: {
+        configSets: mapValues(this._configSets, configNames => configNames.filter(name => nonEmptyConfigs[name] !== undefined)),
+        ...mapValues(configNameToBindResult, c => c.config),
+      },
+      authData: Object.values(configNameToBindResult).map(c => c.authentication).reduce(deepMerge, undefined),
+    };
+  }
+
+}
+
+/**
+ * Options for attach a CloudFormationInit to a resource
+ */
+export interface AttachInitOptions {
+  /**
+   * Instance role of the consuming instance or fleet
+   */
+  readonly instanceRole: iam.IRole;
+
+  /**
+   * OS Platfrom the init config will be used for
+   */
+  readonly platform: InitPlatform;
+
+  /**
+   * UserData to add commands to
+   */
+  readonly userData: UserData;
+
+  /**
+   * ConfigSet to activate
+   *
+   * @default ['default']
+   */
+  readonly configSets?: string[];
+
+  /**
+   * Whether to embed a hash into the userData
+   *
+   * If `true` (the default), a hash of the config will be embedded into the
+   * UserData, so that if the config changes, the UserData changes and
+   * the instance will be replaced.
+   *
+   * If `false`, no such hash will be embedded, and if the CloudFormation Init
+   * config changes nothing will happen to the running instance.
+   *
+   * @default true
+   */
+  readonly embedFingerprint?: boolean;
```

Comment:
Would this be better named with the behaviour rather than the implementation?

Something like `restartOnChange`?

Reply:
I'm open to it, but as a counter-argument: this is one of those where I'd almost rather be slightly implementation-specific and motivate people to read the docs. See the comment here (and on instance.ts); the behavior is nuanced based on whether it's an instance-store backed instance, EBS-backed instance, or part of an ASG (with an UpdatePolicy). Depending on the above, it might be restart _or_ replace, and for some users this distinction may matter.